server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/BluetoothServerCallback.java

```diff
@@ -0,0 +1,217 @@
+package com.google.heartrate.wearos.app.bluetooth.server;
+
+import android.bluetooth.BluetoothDevice;
+import android.bluetooth.BluetoothGatt;
+import android.bluetooth.BluetoothGattCharacteristic;
+import android.bluetooth.BluetoothGattDescriptor;
+import android.bluetooth.BluetoothGattServerCallback;
+import android.bluetooth.BluetoothGattService;
+import android.util.Log;
+
+import com.google.heartrate.wearos.app.bluetooth.server.handlers.GattServiceRequestHandler;
+import com.google.heartrate.wearos.app.gatt.GattException;
+
+import java.util.Arrays;
+import java.util.Set;
+
+/**
+ * Class is used to implement {@link BluetoothServer} callbacks.
+ * <p>General-purpose for all read/write services attributes requests.
+ * Redirects all request to special handler for each service.
+ */
+public class BluetoothServerCallback extends BluetoothGattServerCallback {
+    private static final String TAG = BluetoothServerCallback.class.getSimpleName();
+
+    /** {@link BluetoothServer} to handle requests received from callback. */
+    public final BluetoothServer mBluetoothServer;
+
+    public BluetoothServerCallback(BluetoothServer bluetoothServer) {
+        mBluetoothServer = bluetoothServer;
+    }
+
+    /**
+     * {@link BluetoothGattServerCallback#onConnectionStateChange}
+     *
+     * <p>Notify all request handlers in service about received devise connection state.
+     */
+    @Override
+    public void onConnectionStateChange(BluetoothDevice device, final int status, int newState) {
+        Log.v(TAG, String.format("onConnectionStateChange() - device=%s status=%s state=%s",
+                device.getAddress(), status, newState));
+
+        if (status == BluetoothGatt.GATT_SUCCESS) {
```

Comment:
It might be more readable to fail first to avoid deep `if`
```
if (status != BluetoothGatt.GATT_SUCCESS) {
 log...
 return
}

if (newState == BluetoothGatt.STATE_CONNECTED) 
[...]
```